vowpalwabbit/v_array.h

```diff
@@ -37,36 +40,41 @@ struct v_array
   inline T*& begin() { return _begin; }
   inline T*& end() { return _end; }
 
-  inline T* cbegin() const { return _begin; }
-  inline T* cend() const { return _end; }
+  inline T* const cbegin() const { return _begin; }
+  inline T* const cend() const { return _end; }
 
   // v_array cannot have a user-defined constructor, because it participates in various unions.
   // union members cannot have user-defined constructors.
   // v_array() : _begin(nullptr), _end(nullptr), end_array(nullptr), erase_count(0) {}
   // ~v_array() {
   //  delete_v();
   // }
-  T last() const { return *(_end - 1); }
-  T pop() { return *(--_end); }
-  bool empty() const { return _begin == _end; }
-  void decr() { _end--; }
+  T last() const { return *(_end-1);}
+  T pop() { return *(--_end);}
+  bool empty() const { return _begin == _end;}
+  void decr() { _end--;}
   void incr()
   {
     if (_end == end_array)
       resize(2 * (end_array - _begin) + 3);
     _end++;
   }
   T& operator[](size_t i) const { return _begin[i]; }
+  T& get(size_t i) const { return _begin[i]; }
   inline size_t size() const { return _end - _begin; }
   void resize(size_t length)
-  {
-    if ((size_t)(end_array - _begin) != length)
-    {
-      size_t old_len = _end - _begin;
-      T* temp = (T*)realloc(_begin, sizeof(T) * length);
-      if ((temp == nullptr) && ((sizeof(T) * length) > 0))
-      {
-        THROW("realloc of " << length << " failed in resize().  out of memory?");
+  { 
+  	if ((size_t)(end_array-_begin) != length)
+    { 
+      size_t old_len = _end-_begin;
+      T* temp = (T *)realloc(_begin, sizeof(T) * length);
+      if ((temp == nullptr) && ((sizeof(T)*length) > 0))
+      { 
+#ifdef VW_NOEXCEPT
```

Comment:
Can these operations be made more succinct by embedding them into the THROW macro?

Reply:
Do you mean the #ifdef vw_noexcept and return?   In that case THROW won't work if non-void return functions.